p2p/cuprate-p2p/src/block_downloader.rs

```diff
@@ -0,0 +1,1181 @@
+
+                    self.check_pending_peers(chain_tracker).await;
+
+                    return Ok(());
+                };
+
+                // If the batch is higher than the last time we updated `amount_of_blocks_to_request`, update it
+                // again.
+                if start_height > self.amount_of_blocks_to_request_updated_at {
+                    self.amount_of_blocks_to_request = calculate_next_block_batch_size(
+                        block_batch.size,
+                        block_batch.blocks.len(),
+                        self.config.target_batch_size,
+                    );
+
+                    tracing::debug!(
+                        "Updating batch size of new batches, new size: {}",
+                        self.amount_of_blocks_to_request
+                    );
+
+                    self.amount_of_blocks_to_request_updated_at = start_height;
+                }
+
+                // Add the batch to the queue of ready batches.
+                self.ready_batches_size += block_batch.size;
+                self.ready_batches.push(ReadyQueueBatch {
+                    start_height,
+                    block_batch,
+                });
+
+                // Attempt to push new batches to the buffer.
+                self.push_new_blocks().await?;
+
+                self.pending_peers
+                    .entry(client.info.pruning_seed)
+                    .or_default()
+                    .push(client);
+
+                self.check_pending_peers(chain_tracker).await;
+
+                Ok(())
+            }
+        }
+    }
+
+    /// Starts the main loop of the block downloader.
+    async fn run(mut self) -> Result<(), BlockDownloadError> {
+        let mut chain_tracker = initial_chain_search(
+            &self.client_pool,
+            self.peer_sync_svc.clone(),
+            &mut self.our_chain_svc,
+        )
+        .await?;
+
+        tracing::info!("Attempting to download blocks from peers, this may take a while.");
+
+        let mut check_client_pool_interval = interval(self.config.check_client_pool_interval);
+        check_client_pool_interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        self.check_for_free_clients(&mut chain_tracker).await?;
+
+        loop {
+            tokio::select! {
+                _ = check_client_pool_interval.tick() => {
+                    tracing::debug!("Checking client pool for free peers, timer fired.");
+                    self.check_for_free_clients(&mut chain_tracker).await?;
+
+                     // If we have no inflight requests, and we have had too many empty chain entries in a row assume the top has been found.
+                    if self.inflight_requests.is_empty() && self.amount_of_empty_chain_entries >= EMPTY_CHAIN_ENTRIES_BEFORE_TOP_ASSUMED {
+                        tracing::debug!("Failed to find any more chain entries, probably fround the top");
+                        return Ok(())
```

Comment:
more of a nit on this one though
```suggestion
                        return Ok(());
```